### DISCUSSION TITLE
Can I do MagickNET initialization on startup instead of at first call to the library?

### DISCUSSION THREAD

**Discussion by Nereuzz** [dlemstra/Magick.NET#1043] [Help]
Hi Dirk,

I currently working on an application that uses the `Montage` functionality of MagickNET (great work btw). I was wondering if there is a way to do the MagickNET initialization _before_ I make the first call to `Montage` on a MagickImageCollection? Or is the only workaround to do some "dummy-call" to `Montage` at app startup?

Best regards,
Peter

**Comment by dlemstra** [comment]
You can find information about the initialization here: https://github.com/dlemstra/Magick.NET/tree/main/docs.

**Reply by Nereuzz** [reply]
Thanks for the reply. Unfortunately, I have been trying all of the methods referred to in the documentation, but without luck.

I have tried to make the call `MagickNET.Initialize();` in the constructor of the class where I use `MagickNET` (And before even starting the application), but when creating the class, the initialization call takes a fraction of a second and once the program reaches the call to `collection.Montage(...)`, the first call still takes 15-20 seconds.

I have also tried to both add `var _ = OpenCL.Devices` and set `OpenCL.IsEnables = false` to see if this could be the cause of the problem. I'm running out of ideas, maybe I have misunderstood something? Do you have any other suggestions?

**Reply by dlemstra** [reply]
`OpenCL` is disabled by default in the most recent version. So I wonder what is causing the initial delay. You could use the `MagickNET.Log` and `MagickNET.SetLogEvents` methods to get detailed logging information (more info see the docs) that will probably show where the delay is. And what is the exact version that you are using?

**Reply by Nereuzz** [reply]
I will try using more detailed logs to see if I can find the delay. I'm using `Magick.NET-Q16-AnyCPU` version `8.3.2`

**Reply by Nereuzz** [reply]
After a bit of digging in the logs I have found the position of the delay:

```
{"Message":"2021-10-04T14:28:26+00:00 0:12.228 1.200u 7.1.0 Configure testMagick[]: type.c/LoadTypeCache/1085/Configure\n  Loading type configure file \".config/ImageMagick/type-ghostscript.xml\
Fontconfig error: Cannot load default config file
{"Message":"2021-10-04T14:29:06+00:00 0:52.364 41.400u 7.1.0 Annotate testMagick[]: annotate.c/GetTypeMetrics/887/Annotate\n  Metrics: text: ; width: 0; height: 13; ascent: 10; descent: -3; max advance: 12; bounds: 0,-3  7,7; origin: 0,0; pixels per em: 12,12; underline position: -1.5; underline thickness: 0.597656"}
```

Looking at the timestamps, there are 30 seconds between the two events around the `Fontconfig` error. Could this be the cause?
I am manually setting a Font in the `MontageSettings` struct, which I know is available to `MagickNET`.

And the above logs happen when the call to `collection.Montage(images)` is made, not during `MagickNET.Initialize()`.

**Reply by dlemstra** [reply]
It does look like the call to `LoadFontConfigFonts` inside ImageMagick is causing the delay. Not sure what that would need to take around 40 seconds. I think that you can force the initial loading by doing the following: `MagickNET.FontNames.ToList()`.

**Reply by Nereuzz** [reply]
Still not sure why the call to `LoadFontConfigFonts` takes 30-40 seconds, but I managed to move the delay to the app startup phase by using your suggested `MagickNET.FontNames.ToList()` call. Thanks a lot!

I will try to dig up more information on why the call to `LoadFontConfigFonts` is slow.

**Reply by dlemstra** [reply]
Can you reproduce this issue inside a container? That will make it easier for me to debug the issue.

**Reply by Nereuzz** [reply]
Sorry for the late answer, haven't really had time to code much the past days.

I will try to reproduce the issue in a container during the weekend.
